win32ss/gdi/gdi32/objects/bitmap.c

```diff
@@ -677,6 +681,62 @@ SetDIBitsToDevice(
     if (!pConvertedInfo)
         return 0;
 
+    if (pConvertedInfo->bmiHeader.biHeight == 0)
+    {
+        return 0;
+    }
+
+    if (!aligned(Bits, 4) && (ScanLines > Height))
+    {
+        return 0;
+    }
+
+    /* Below code modeled after Wine's nulldrv_SetDIBitsToDevice */
+    bmiHeight = abs(pConvertedInfo->bmiHeader.biHeight);
+    if (StartScan <= YSrc + bmiHeight)
+    {
+        BOOL top_down;
+        UINT src_y = 0;
+
+        top_down = (pConvertedInfo->bmiHeader.biHeight < 0);
+        if ((pConvertedInfo->bmiHeader.biCompression == BI_RLE8) ||
+            (pConvertedInfo->bmiHeader.biCompression == BI_RLE4))
+        {
+            StartScan = 0;
+            ScanLines = bmiHeight;
+        }
+        else
+        {
+            if (StartScan >= bmiHeight)
+            {
+                return 0;
+            }
+            if (!top_down && ScanLines > bmiHeight - StartScan)
+            {
+                ScanLines = bmiHeight - StartScan;
+            }
+            src_y = StartScan + ScanLines - (YSrc + Height);
+            if (src_y > 0)
+            {
+                if (!top_down)
+                {
+                    /* get rid of unnecessary lines */
+                    if (src_y >= ScanLines)
+                    {
+                     return 0;
+                    }
+                    ScanLines -= src_y;
+                }
+                else if (src_y >= ScanLines)
+                {
+                    return ScanLines;
+                }
+            }
+            LinesCopied = ScanLines;
+        }
+    }
+
+   /* This call is totally non-functional and can be removed */
```

Comment:
Well, what would happen for Metafiles support then?

Reply:
@HBelusca in https://jira.reactos.org/browse/CORE-12377, I posted a test patch
https://jira.reactos.org/secure/attachment/65525/objects_bitmap_test.patch
that just patches out this section of code and then ran the testbots on it. The results are as follows:

objects_bitmap_test.patch JID65525 on top of 0.4.15-dev-5905-gce56272

VBox: https://reactos.org/testman/compare.php?ids=86721,86724 LGTM
KVM:  https://reactos.org/testman/compare.php?ids=86720,86725 LGTM

Based on what I see here. there are no substantial differences. Maybe because there are no tests? I do not understand what this code is supposed to do anyway. Maybe one of the developers can propose some test procedure for this? Thanks.

Reply:
The tests don't appear to exercise metafiles fully, so please don't modify more than what's needed.

Reply:
I modified nothing in regard to this. I only added a comment. I will remove the comment. Thanks.